internal/proxy/cache_test.go

```diff
@@ -0,0 +1,498 @@
+			assert.Equal(t, tc.expected, shallFetch)
+		})
+	}
+}
+
+func TestShallRefreshCache(t *testing.T) {
+	tests := map[string]struct {
+		cache        *Cache
+		cacheControl string
+		expected     bool
+	}{
+		"Shall refresh cache when cache-control header is 'cache'": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second}),
+			cacheControl: headerOptionCacheable,
+			expected:     true,
+		},
+		"Shall refresh cache when cache-control header is 'recache'": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second}),
+			cacheControl: headerOptionRecache,
+			expected:     true,
+		},
+		"Shall refresh cache when cache-control header is 'cache,recache'": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second}),
+			cacheControl: strings.Join([]string{headerOptionCacheable, headerOptionRecache}, ","),
+			expected:     true,
+		},
+		"Shall not refresh cache when cache is nil": {
+			cache:        nil,
+			cacheControl: headerOptionCacheable,
+			expected:     false,
+		},
+		"Shall not refresh cache when cache-control header is incorrect": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second}),
+			cacheControl: "nope",
+			expected:     false,
+		},
+		"Shall not refresh cache when cache-control header is empty": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second}),
+			cacheControl: "",
+			expected:     false,
+		},
+		"No cache when cache-control header is nocache": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second}),
+			cacheControl: headerOptionNocache,
+			expected:     false,
+		},
+		"Shall refresh cache when default enabled": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second, DefaultEnabled: true}),
+			cacheControl: "",
+			expected:     true,
+		},
+		"Shall not refresh cache when default enabled and 'nocache' set": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second, DefaultEnabled: true}),
+			cacheControl: headerOptionNocache,
+			expected:     false,
+		},
+		"Shall refresh cache when default enabled and 'recache' set": {
+			cache:        NewCache(config.CacheConfig{TTL: 10 * time.Second, DefaultEnabled: true}),
+			cacheControl: headerOptionRecache,
+			expected:     true,
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			r := httptest.NewRequest("GET", "/v1/uuid/s1", nil)
+			r.Header.Set(VaultCacheControlHeaderName, tc.cacheControl)
+			shallFetch := shallRefreshCache(r, tc.cache)
+			assert.Equal(t, tc.expected, shallFetch)
+		})
+	}
+}
+
+func TestRetrieveData(t *testing.T) {
+	r := httptest.NewRequest("GET", "/v1/uuid/s1", nil)
+	r.Header.Set(VaultCacheControlHeaderName, headerOptionCacheable)
+	statusCode := 200
+	body := "Hello World"
+	resp := &http.Response{
+		Status:        "200 OK",
+		StatusCode:    statusCode,
+		Proto:         "HTTP/1.1",
+		ProtoMajor:    1,
+		ProtoMinor:    1,
+		Body:          ioutil.NopCloser(bytes.NewBufferString(body)),
```

Comment:
nit: `retrieveData` shouldn't touch the request body, so it would be a slightly tighter test if this is nil or a different body.

Reply:
Done in bd90b345c483bfb20438482fa489a3ef54380bfc.